jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -162,28 +201,42 @@ impl From<blockcfg::BlockDate> for BlockDate {
     }
 }
 
+#[derive(Clone)]
 struct Transaction {
     id: FragmentId,
-    in_block: HeaderHash,
+    in_block: Option<HeaderHash>,
 }
 
 impl Transaction {
     fn from_id(id: FragmentId, context: &Context) -> FieldResult<Transaction> {
-        let in_block =
-            context
-                .db
-                .find_block_by_transaction(&id)
-                .wait()?
-                .ok_or(ErrorKind::NotFound(format!(
-                    "transaction not found: {}",
-                    &id,
-                )))?;
+        let in_block = Self::get_in_block(&id, context).ok_or(ErrorKind::NotFound(format!(
+            "transaction not found: {}",
+            &id,
+        )))?;
 
-        Ok(Transaction { id, in_block })
+        Ok(Transaction {
+            id,
+            in_block: Some(in_block),
+        })
+    }
+
+    fn from_valid_id(id: FragmentId) -> Transaction {
+        Transaction { id, in_block: None }
+    }
+
+    fn get_in_block(id: &HeaderHash, context: &Context) -> Option<HeaderHash> {
```

Comment:
This has no relation to the `Transaction` type, so it's better be a method on `Context` or, if you don't define that (can't quickly find where it's coming from), as a free function.